compiler-plugin/src/main/java/io/ballerina/stdlib/grpc/plugin/GrpcServiceValidator.java

```diff
@@ -189,64 +189,68 @@ private void validateServiceFunctions(FunctionDefinitionNode functionDefinitionN
     private void validateFunctionSignature(FunctionDefinitionNode functionDefinitionNode,
                                            SyntaxNodeAnalysisContext syntaxNodeAnalysisContext, String serviceName) {
 
-        FunctionSignatureNode functionSignatureNode = functionDefinitionNode.functionSignature();
-        SeparatedNodeList<ParameterNode> parameterNodes = functionSignatureNode.parameters();
-
-        if (parameterNodes.size() == 2) {
-            String firstParameter = typeNameFromParameterNode(functionSignatureNode.parameters().get(0));
-
-            if (!firstParameter.toLowerCase(Locale.ENGLISH).contains(GRPC_GENERIC_CALLER)) {
-                reportErrorDiagnostic(functionDefinitionNode, syntaxNodeAnalysisContext,
-                        GrpcCompilerPluginConstants.CompilationErrors.TWO_PARAMS_WITHOUT_CALLER.getError(),
-                        GrpcCompilerPluginConstants.CompilationErrors.TWO_PARAMS_WITHOUT_CALLER.getErrorCode());
-            } else if (!(firstParameter.startsWith(serviceName) && firstParameter.endsWith(GRPC_EXACT_CALLER))) {
-                String expectedCaller = serviceName + GRPC_RETURN_TYPE + GRPC_EXACT_CALLER;
-                String diagnosticMessage = GrpcCompilerPluginConstants.CompilationErrors
-                        .INVALID_CALLER_TYPE.getError() +
-                        expectedCaller + "\" but found \"" + firstParameter + "\"";
-                reportErrorDiagnostic(functionDefinitionNode, syntaxNodeAnalysisContext, diagnosticMessage,
-                        GrpcCompilerPluginConstants.CompilationErrors.INVALID_CALLER_TYPE.getErrorCode());
-            } else if (functionSignatureNode.returnTypeDesc().isPresent()) {
-                Node returnTypeNode = functionSignatureNode.returnTypeDesc().get().type();
-                if (!(SyntaxKind.NIL_TYPE_DESC == returnTypeNode.kind())) {
-                    if (SyntaxKind.OPTIONAL_TYPE_DESC == returnTypeNode.kind() &&
-                            ((OptionalTypeDescriptorNode) returnTypeNode).children().size() == 2) {
-                        for (Node value : ((OptionalTypeDescriptorNode) returnTypeNode).children()) {
-                            if (!(value.kind() == SyntaxKind.ERROR_TYPE_DESC ||
-                                    value.kind() == SyntaxKind.QUESTION_MARK_TOKEN)) {
-                                reportErrorDiagnostic(functionDefinitionNode, syntaxNodeAnalysisContext,
-                                        GrpcCompilerPluginConstants.CompilationErrors.RETURN_WITH_CALLER.getError(),
-                                        GrpcCompilerPluginConstants.CompilationErrors
-                                                .RETURN_WITH_CALLER.getErrorCode());
-                                break;
+        boolean isRemoteFunction = functionDefinitionNode.qualifierList().stream()
+                .filter(q -> q.kind() == SyntaxKind.REMOTE_KEYWORD).toArray().length == 1;
+
+        if (isRemoteFunction) {
```

Comment:
```suggestion
        if (functionDefinitionNode.qualifierList().stream()
                .filter(q -> q.kind() == SyntaxKind.REMOTE_KEYWORD).toArray().length == 1) {
```

Reply:
Since this variable isn't reused.

Reply:
I thought having the `isRemoteFunction` improve the readability since this is a nested if-else block.

Reply:
what if we do this check when calling this function